pkg/archive/archive.go

```diff
@@ -856,179 +856,189 @@ func Tar(path string, compression Compression) (io.ReadCloser, error) {
+			if seen[relFilePath] {
+				return nil
+			}
+			seen[relFilePath] = true
+
+			// Rename the base resource.
+			if rebaseName != "" {
+				var replacement string
+				if rebaseName != string(filepath.Separator) {
+					// Special case the root directory to replace with an
+					// empty string instead so that we don't end up with
+					// double slashes in the paths.
+					replacement = rebaseName
+				}
 
-					for _, pat := range pm.Patterns() {
-						if !pat.Exclusion() {
-							continue
-						}
-						if strings.HasPrefix(pat.String()+string(filepath.Separator), dirSlash) {
-							// found a match - so can't skip this dir
-							return nil
-						}
-					}
+				relFilePath = strings.Replace(relFilePath, include, replacement, 1)
+			}
 
-					// No matching exclusion dir so just skip dir
-					return filepath.SkipDir
+			if err := ta.addFile(filePath, relFilePath); err != nil {
+				logrus.Errorf("Can't add file %s to tar: %s", filePath, err)
+				// if pipe is broken, stop writing tar stream to it
+				if err == io.ErrClosedPipe {
+					return err
 				}
+			}
+			return nil
+		}); err != nil {
+			return err
+		}
+	}
+	return ta.TarWriter.Close()
+}
 
-				if seen[relFilePath] {
-					return nil
-				}
-				seen[relFilePath] = true
-
-				// Rename the base resource.
-				if rebaseName != "" {
-					var replacement string
-					if rebaseName != string(filepath.Separator) {
-						// Special case the root directory to replace with an
-						// empty string instead so that we don't end up with
-						// double slashes in the paths.
-						replacement = rebaseName
-					}
+// TarWithOptions creates an archive from the directory at `srcPath`, only including files whose relative
+// paths are included in `options.IncludeFiles` (if non-nil) or not in `options.ExcludePatterns`.
+func TarWithOptions(srcPath string, options *TarOptions) (io.ReadCloser, error) {
+	if _, err := fileutils.NewPatternMatcher(options.ExcludePatterns); err != nil {
+		return nil, err
+	}
 
-					relFilePath = strings.Replace(relFilePath, include, replacement, 1)
-				}
+	pipeReader, pipeWriter := io.Pipe()
 
-				if err := ta.addFile(filePath, relFilePath); err != nil {
-					logrus.Errorf("Can't add file %s to tar: %s", filePath, err)
-					// if pipe is broken, stop writing tar stream to it
-					if err == io.ErrClosedPipe {
-						return err
-					}
-				}
-				return nil
-			}); err != nil {
-				logrus.Errorf("%s", err)
-				return
-			}
+	compressWriter, err := CompressStream(pipeWriter, options.Compression)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		if err := TarWithOptionsTo(compressWriter, srcPath, options); err != nil {
+			logrus.Errorf("Can't create tar archive: %s", err)
+		}
+		if err := compressWriter.Close(); err != nil {
+			logrus.Errorf("Can't close compress writer: %s", err)
+		}
+		if err := pipeWriter.Close(); err != nil {
+			logrus.Errorf("Can't close pipe writer: %s", err)
 		}
```

Comment:
Moving these things out of `defer` is incorrect, panics will now cause the consuming goroutine to hang.

I think the right pattern for a pipe-producing goroutine is to:
- Split as much as possible into an “ordinary” function.
- Add a single-purpose goroutine function whose _only purpose_ is to _guarantee_ error handling, i.e. that the goroutine will exit, that the consuming goroutine will not block, and that errors, if any, are correctly reported.

Compare https://github.com/containers/image/blob/e4a0c90bdc94a22db79aad8d278c203819ffab42/copy/compression.go#L414 .

Reply:
>  “ordinary” function.

By this I mean a function that returns `error` the usual way, without having to use some other error reporting/propagating mechanism. Then that unusual error reporting mechanism only exists in _one_ easily-verifiable place and users who add new error paths to the large function don’t risk forgetting to do something unusual.

Reply:
c/storage is a library, and used from other projects; Podman does seem to have a panic handler, at least at the top level.

But you are right that that does not prevent this function from crashing the whole process.